lib/application/widgets/theme/custom_button.dart

```diff
@@ -0,0 +1,313 @@
+
+  /// Builds the button background color based on the button state.
+  final Color? Function(ButtonState state) backgroundColorBuilder;
+
+  @override
+  Widget build(BuildContext context) {
+    final textTheme = Theme.of(context).textTheme.button!;
+
+    Widget leadingAssetBuilder(ButtonState state) => Image.asset(leadingAsset!, color: textTheme.color);
+
+    return CustomButton(
+      text: text,
+      onPressed: onPressed,
+      isPressedOverlayEnabled: true,
+      backgroundColorBuilder: backgroundColorBuilder,
+      textStyleBuilder: (state) => textTheme,
+      leadingWidgetBuilder: leadingAsset != null ? leadingAssetBuilder : null,
+    );
+  }
+}
+
+/// Shared visual style between all custom button's.
+///
+/// Provides a common and customizable button style that must be used by all custom buttons.
+///
+/// See also:
+///
+///   * [PrimaryElevatedButton] - The primary visually opinionated alternative to [ElevatedButton].
+///   * [SecondaryElevatedButton] - The secondary visually opinionated alternative to [ElevatedButton].
+///   * [CustomTextButton] - The visually opinionated alternative to [TextButton].
+@visibleForTesting
+class CustomButton extends StatefulWidget {
+  const CustomButton({
+    required this.text,
+    this.onPressed,
+    this.isPressedOverlayEnabled = false,
+    this.leadingWidgetBuilder,
+    this.borderBuilder,
+    this.backgroundColorBuilder,
+    this.backgroundGradientBuilder,
+    this.textStyleBuilder,
+  });
+
+  /// The button text.
+  ///
+  /// Its style may be customized by implementing [textStyleBuilder].
+  final String text;
+
+  final VoidCallback? onPressed;
+
+  /// If `true` adds a overlay over the button when it's pressed.
+  final bool isPressedOverlayEnabled;
+
+  /// Builds the button leading widget based on the button `state`.
+  final Widget Function(ButtonState state)? leadingWidgetBuilder;
+
+  /// Builds the button border based on the button `state`.
+  ///
+  /// If `null` no border is added.
+  final BoxBorder? Function(ButtonState state)? borderBuilder;
+
+  /// Builds the button background color based on the button `state`.
+  ///
+  /// Will be overriden by [backgroundGradientBuilder] if implemented.
+  /// If `null` no background color is added.
+  final Color? Function(ButtonState state)? backgroundColorBuilder;
+
+  /// Builds the button background gradient based on the button `state`.
+  ///
+  /// Overrides [backgroundColorBuilder] if both are implemented.
+  /// If `null` no background gradient is added.
+  final LinearGradient? Function(ButtonState state)? backgroundGradientBuilder;
+
+  /// Builds the button [text] style based on the button `state`.
+  ///
+  /// If `null` the default text theme is used.
+  final TextStyle? Function(ButtonState state)? textStyleBuilder;
+
+  @override
+  CustomButtonState createState() => CustomButtonState();
+}
+
+@visibleForTesting
+class CustomButtonState extends State<CustomButton> {
+  late ButtonState state;
```

Comment:
no need to create a stateful widget, you can use a stateless widget with a state hook